### PR TITLE
Initial version

### PR DESCRIPTION
**Pull request by zrzka**
It's an initial version, still prototype. I've decided to merge it to the master and create smaller PRs for future changes like documentation, optimisation, ...

### CODE REVIEW

**Comment by zrzka** [comment]
@cyplo TIA :)

File: src/parser/mod.rs [hunk 1]
```diff
@@ -0,0 +1,383 @@
+use crate::{
```
Comment (cyplo) [inline src/parser/mod.rs]: the whole file seems quite readable, albeit a bit long - would it make sense to split into multiple files later maybe ?

File: src/parser/mod.rs [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+//
+// square_brackets = _{
+//     "[" ~ (integer | string | dotted_square_bracket_identifier) ~ "]"
+// }
+//
+// dotted_square_bracket_identifier = ${
+//     identifier ~ ( ("." ~ identifier) | square_brackets )*
+// }
+//
+fn _parse_dotted_square_bracket_identifier(pair: Pair<Rule>) -> Result<Identifier> {
```
Comment (cyplo) [inline src/parser/mod.rs]: It may no be immediately obvious why the function name starts with underscore - I was thinking for few seconds that this has to do with the `unused` convention in Rust. My current understanding is that the result is wrapped in the function `parse_dotted_square_bracket_identifier` (without the underscore in the name) but also this one calls itself recursively - which makes this difficult to name, I agree. Just leaving a note here without much suggestion tbh - maybe a name change could help, maybe a sentence in the function documentation ?

File: src/builtin/filter.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    Ok(Value::String(s.to_lowercase()))
+}
+
+pub(crate) fn upper(value: &Value) -> Result<Value> {
+    let s = value.as_str().ok_or_else(|| "`upper` filter accepts string only")?;
+    Ok(Value::String(s.to_uppercase()))
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (cyplo) [inline src/builtin/filter.rs]: I like the level of detail in the tests here and in other places :) What do you think about adding `quickcheck` tests later ? To try finding edge cases automatically ? May be useful for a generic tool like this library.

File: src/engine/context.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
+use crate::error::{bail, Result};
+use crate::parser::ast::*;
+use serde_json::Value;
+
+pub struct Context {
```
Comment (cyplo) [inline src/engine/context.rs]: Is is not immediately clear what the context may be used for just by looking at the name/definition. I gather that probably it's a flying parameter, used to pass evaluation state ? Would a bit of documentation here help ?

File: src/engine/context.rs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                Value::Number(ref x) => x
+                    .as_i64()
+                    .ok_or_else(|| "unable to lookup variable: invalid integer index".into())
+                    .and_then(|idx| self.lookup_variable_value(value, &IdentifierValue::IntegerIndex(idx as isize))),
+                _ => bail!("unable to lookup variable: result of indirect lookup is not a number / string"),
+            },
+        }
+    }
+
+    pub(crate) fn lookup_variable(&self, identifier: &Identifier) -> Result<&Value> {
```
Comment (cyplo) [inline src/engine/context.rs]: nit: may make sense to put the `pub fn`s before the private ones

File: src/engine/context.rs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    data: Value,
+}
+
+impl Context {
+    pub fn new(data: Value) -> Context {
+        Context { data }
+    }
+
+    fn lookup_variable_value<'a>(&self, value: &'a Value, identifier: &IdentifierValue) -> Result<&'a Value> {
+        match identifier {
```
Comment (cyplo) [inline src/engine/context.rs]: This match may benefit from extracting functions per match-arm, it's a bit lengthy for me to be able to read and comprehend at once.

File: src/engine/mod.rs [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+
+pub struct Engine {
+    functions: HashMap<String, FunctionFn>,
+    filters: HashMap<String, FilterFn>,
+}
+
+impl Default for Engine {
+    fn default() -> Engine {
+        let mut filters = HashMap::new();
+        filters.insert("upper".into(), Box::new(filter::upper) as FilterFn);
```
Comment (cyplo) [inline src/engine/mod.rs]: Just thinking out loud here, no need to do anything :
Would it be possible (and would it make sense?) to do the registration of filters and other functions during compilation time, using e.g. macro ?

File: src/engine/mod.rs [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+        Engine { functions, filters }
+    }
+}
+
+impl Engine {
+    fn eval_math(&self, lhs: &Number, rhs: &Number, operator: MathOperator) -> Result<Number> {
+        match operator {
+            MathOperator::Addition => {
+                if lhs.is_i64() && rhs.is_i64() {
+                    if let Some(x) = lhs.as_i64().unwrap().checked_add(rhs.as_i64().unwrap()) {
```
Comment (cyplo) [inline src/engine/mod.rs]: Right now the `if let` skips the `None` variant checking which signifies the integer overflow on addition.
While unlikely - may be worth forwarding the error of the overflow condition to the user here, or even panic'ing. - what do you think ?

File: src/engine/mod.rs [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+                }
+
+                let lhs = lhs.as_f64().unwrap();
+                let rhs = rhs.as_f64().unwrap();
+                let result = lhs + rhs;
+
+                Ok(Number::from_f64(validate_f64(result)?).unwrap())
+            }
+            MathOperator::Subtraction => {
+                if lhs.is_i64() && rhs.is_i64() {
```
Comment (cyplo) [inline src/engine/mod.rs]: the patter of `if..as..unwrap...FUNCTION...as...return` seems to be used throughout this function, would it make sense to refactor to extract a function that takes a lambda ?

File: src/engine/mod.rs [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+                let rhs = self.eval_as_number(rhs, context)?;
+                self.eval_math(&lhs, &rhs, *operator)?
+            }
+            ExpressionValue::FunctionCall(FunctionCall { ref name, ref args }) => {
+                match self.eval_function(name, args, context)? {
+                    Value::Number(x) => x,
+                    _ => bail!("result of `{}` is not a number", name),
+                }
+            }
+            ExpressionValue::Boolean(_) => bail!("unable to evaluate boolean as a number"),
```
Comment (cyplo) [inline src/engine/mod.rs]: Curious - would it be possible to express this in the type system ? E.g. make the function only take things that it can work on, like introducing a 'smaller' type ? Or would that spin into to much work in other places ?

File: src/engine/mod.rs [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+                result = Value::Bool(!x);
+            } else {
+                bail!("unable to negate non bool value");
+            }
+        }
+
+        Ok(result)
+    }
+
+    fn eval_value_as_bool(&self, value: &ExpressionValue, context: &Context) -> Result<bool> {
```
Comment (cyplo) [inline src/engine/mod.rs]: may benefit from extracting smaller functions from the inner match branches

File: tests/engine/eval/lookup.rs [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+fn test_square_bracket_multiple_indirect() {
+    let ctx = Context::new(json!({
+        "country": "Czech Republic",
+        "city": "Hradec Kralove",
+        "data": {
+            "country": {
+                "Czech Republic": {
+                    "city": {
+                        "Hradec Kralove": {
+                            "rust-developers": 1
```
Comment (cyplo) [inline tests/engine/eval/lookup.rs]: I would add something about dobrého vojáka Švejka as well maybe :D

File: .idea/runConfigurations/Test.xml [hunk 13]
```diff
@@ -1,9 +1,9 @@
 <component name="ProjectRunConfigurationManager">
   <configuration default="false" name="Test" type="CargoCommandRunConfiguration" factoryName="Cargo Command">
     <option name="channel" value="DEFAULT" />
-    <option name="command" value="test" />
-    <option name="nocapture" value="true" />
-    <option name="backtrace" value="SHORT" />
+    <option name="command" value="test -- --nocapture" />
+    <option name="allFeatures" value="true" />
+    <option name="backtrace" value="FULL" />
```
[the remark this answers is not in the document]
Comment (zrzka) [reply .idea/runConfigurations/Test.xml]: You can always do this locally. All tests should pass, ideally, you shouldn't see any backtrace at all. And if it fails, it's very useful to have the full backtrace to know where the problem is. IMHO it's not worth to have so many run configurations.

**Comment by cyplo** [comment]
Let me know if you want to merge straight away or prefer to work with PR open- I can click on approve or merge if needed

File: README.md [hunk 14]
```diff
@@ -1 +1,18 @@
...
+## Status
+
+[![Build Status](https://travis-ci.org/balena-io-modules/balena-temen.svg?branch=master)](https://travis-ci.org/balena-io-modules/balena-temen)
+
+Undocumented prototype with lot of issues like unnecessary cloning, etc. Not ready for
+production use.
+
+## Purpose
+
+Main goal of this library is to provide a way how to:
```
[the remark this answers is not in the document]
Comment (zrzka) [reply README.md]: As I wrote, all these nitpicks are part of these [issues](https://github.com/balena-io-modules/balena-temen/issues) and will be resolved later in smaller chunks.

File: rustfmt.toml [hunk 15]
```diff
@@ -0,0 +1,2 @@
+max_width = 120
+comment_width = 120
```
[the remark this answers is not in the document]
Comment (zrzka) [reply rustfmt.toml]: Removed.

File: src/lib.rs [hunk 16]
```diff
@@ -1,7 +1,19 @@
...
-    }
-}
+//! [Balena](https://www.balena.io) TEMplating ENgine
+//!
+//! A crate that provides facilities to:
+//!
+//! * evaluate expressions
+//! * evaluate logical expressions
+//! * retrieve AST of any expression for further validation
+#![allow(clippy::possible_missing_comma)]
```
[the remark this answers is not in the document]
Comment (zrzka) [reply src/lib.rs]: Not worth IMHO. The problem is with fmt preview (defaults), which isn't predictable yet. I mean, it's not stable yet and the code is sometimes formatted with trailing `,` and sometimes not. Thus I'm ignoring this one completely for now.

File: src/parser/grammar.pest [hunk 17]
```diff
@@ -0,0 +1,124 @@
+//
+// https://docs.rs/pest_derive/2.0.1/pest_derive/#modifiers
```
[the remark this answers is not in the document]
Comment (zrzka) [reply src/parser/grammar.pest]: Removed

File: src/parser/mod.rs (hunk 1 above)
Comment (zrzka) [reply src/parser/mod.rs]: 400 hundred lines seems ok to me. I mean, there's still work to do. It can be shorted or longer in the end and then we can split it. I don't like long files as well, just skipping this as it's still under the construction :)

File: src/parser/mod.rs (hunk 2 above)
Comment (zrzka) [reply src/parser/mod.rs]: Renamed, not sure it helps a lot. Anyway, I'll probably rename all of these functions (include expression in the name for example to be more clear what it parses).

File: src/engine/context.rs (hunk 4 above)
Comment (zrzka) [reply src/engine/context.rs]: For now, it's just a _placeholder_ for context data. The plan is to include more things here and it will make it much more clearer. Especially when #9 will be implemented.

File: src/engine/mod.rs (hunk 7 above)
Comment (zrzka) [reply src/engine/mod.rs]: For what purpose? Once #8 will be implemented this will be removed and public registration functions will be used.

File: src/engine/mod.rs (hunk 8 above)
Comment (zrzka) [reply src/engine/mod.rs]: Disagree. I do not want to panic at all. I consider panicking as a bad behavior and no library should panic, especially when we're talking about user input here.

The decission I made here is that I try to do float math if i64 fails and then fail. Don't have strong opinion about this.

File: src/engine/mod.rs (hunk 9 above)
Comment (zrzka) [reply src/engine/mod.rs]: Agree here, but will probably do it later if you don't mind. Did add TODO here.

File: src/engine/mod.rs (hunk 10 above)
Comment (zrzka) [reply src/engine/mod.rs]: It will, but the AST will be much more complicated and there will be lot of other work to do. Which isn't worth IMHO. Especially now when I still do expect grammar changes.

File: tests/engine/eval/lookup.rs (hunk 12 above)
Comment (zrzka) [reply tests/engine/eval/lookup.rs]: :D

**Comment by zrzka** [comment]
@cyplo looking for approval here :) I did respond to all comments, added TODO, made small fixes, ... Rest is [here](https://github.com/balena-io-modules/balena-temen/issues). There will be another round of PRs and final, thorough, review before release.

**Comment by cyplo** [comment]
go for it - may have more comments after using it but otherwise feel free to merge at will @zrzka
